text/0000-release-meta-spec-v2.md

```diff
@@ -0,0 +1,464 @@
+
+# RFC--0 --- PGXN Release Meta Spec v2
+
+## Abstract
+
+This document describes version 2.0.0 of the *release* format for [PGXN]
+source distribution metadata. It extends the *distribution* [PGXN Meta Spec
+v2][v2] `META.json` provided by authors in PGXN uploads PGXN with signed
+metadata about the release on PGXN. This will allow clients to verify the
+validity of PGXN release sources via public key verification.
+
+## Introduction
+
+### Background
+
+When [PGXN Manager] was implemented in 2010, it published release zip files
+that contain author-supplied [PGXN Meta Spec v1][v1] metadata, and also
+published an augmented `META.json` file with three additional properties
+representing the release:
+
+*    `user`: The name of the user who made the release
+*    `date`: A timestamp for the release
+*    `sha1`: A [SHA-1] digest of the release zip file
+
+Clients were advised to download both this release `META.json` along with the
+source zip file, and to validate the zip file against the checksum. The `user`
+and `date` information were provided mainly for informational purposes.
+
+Compare, for example, the pair-0.1.7 [release META.json] to the
+author-provided [distribution META.json]. The difference is these three
+fields:
+
+``` json
+{
+  "user": "theory",
+  "date": "2020-10-25T21:54:02Z",
+  "sha1": "5b9e3ba948b18703227e4dea17696c0f1d971759"
+}
+```
+
+Using the release metadata, a client can determine the URL to download the zip
+file, then validate it against the SHA-1 digest.
+
+### Signed Releases
+
+A lot has changed since 2010, including an increasing need for [public key
+signing] to validate that distribution files come from trusted sources. [PGXN
+Meta Spec v2][v2] provides an opportunity to update the release `META.json`
+format with signed metadata to enable a much more secure method of validation.
+
+This RFC therefore proposes to extend [v2] distribution metadata with a single
+additional property, `release`, that contains a [JWS JSON Serialization]
+object as defined by [RFC 7515][JWS]. This property will allow clients not
+only to find the release file to download and verify against checksums, but
+also validate it against a public key provided by PGXN.
+
+The design allows multiple digital signatures, which in the future may allow
+authors or other entities to sign releases with their own keys. The new format
+would append a structure such as this to the distribution `META.json` file:
+
+``` json
+{
+  "release": {
+     "pgxn": {
+       "payload": "eyJ1c2VyIjoidGhlb3J5IiwiZGF0ZSI6IjIwMjQtMDktMTNUMTc6MzI6NTVaIiwidXJpIjoiZGlzdC9wYWlyLzAuMS43L3BhaXItMC4xLjcuemlwIiwiZGlnZXN0cyI6eyJzaGE1MTIiOiJiMzUzYjVhODJiM2I1NGU5NWY0YTI4NTllN2EyYmQwNjQ4YWJjYjM1YTdjMzYxMmIxMjZjMmM3NTQzOGZjMmY4ZThlZTFmMTllNjFmMzBmYTU0ZDdiYjY0YmNmMjE3ZWQxMjY0NzIyYjQ5N2JjYjYxM2Y4MmQ3ODc1MTUxNWI2NyJ9fQ",
+       "signatures": [
+          {
+            "protected":"eyJhbGciOiJSUzI1NiJ9",
+            "header": {"kid": "2024-12-29" },
```

Comment:
Normally a JWS is like `headerB64.payloadB64.signatureB64`, and therefore I am wondering about why the payload is separate from the header and signature. I believe it is like that in order to support multiple signatures.

Questions to consider are...
- is it necessary to support multiple signatures?
- For 'protected' and 'header', both of these parts typically included in the "header" as described in RFC 7515, so why not combine that into just 'header' as b64 encoded.
- Wondering because if you take the payload, the header (as suggested in previous comment), signature, and concatenate them with "." joined ([RFC7515 section 3.3](https://datatracker.ietf.org/doc/html/rfc7515#section-3.3)), then would that make it exactly like a "normal" signature? i.e. easier to use with existing clients that do signature validation maybe.

Reply:
It's an official RFC 7515 format called [JWS JSON Serialization](https://www.rfc-editor.org/rfc/rfc7515.html#section-7.2). We don't need URL-safe serialization, so this is just a little friendlier for clients. And yes, it allows multiple signatures, which could be useful for author signing or gradual key rotation.

1. I think it's useful to support multiple signatures because, unlike ephemeral HTTP requests, there's are records at rest and will be around for many years.
2. `protected` and `header` are not combined because the [JWS JSON Serialization](https://www.rfc-editor.org/rfc/rfc7515.html#section-7.2) format separates them.
3. Clients should also support this format, as it's part of the standard. But even if they don't, conversion will be trivial, as you point out.

Really I wanted the payload not to be Base64 URL-encoded, as specified in [JWS-JS](https://datatracker.ietf.org/doc/html/draft-jones-json-web-signature-json-serialization-01), but I believe that spec was abandoned in favor of [RFC 7515 JWS JSON Serialization](https://www.rfc-editor.org/rfc/rfc7515.html#section-7.2), so going with that approved standard, instead.